File: german-telework/service/gateway/actions/leave.py
import actions.user

class Leave:

    def __init__(self, ident, request_type, start_date, end_date, reason, destination, emergency_phone): 
        self.ident = ident
        self.request_type = request_type
        self.start_date = start_date
        self.end_date = end_date
        self.reason = reason
        self.destination = destination
        self.emergency_phone = emergency_phone


    def serialize(self):
        res = ""
        res += self.ident
        res += "|" + self.request_type
        res += "|" + self.start_date 
        res += "|" + self.end_date
        res += "|" + self.reason
        res += "|" + self.destination
        res += "|" + self.emergency_phone
        return res


    def deserialize(data):
        parts = data.split("|")
        ident = parts[0]
        request_type = parts[1]
        start_date = parts[2]
        end_date = parts[3]
        reason = parts[4]
        destination = parts[5]
        emergency_phone = parts[6]
        return Leave(ident, request_type, start_date, end_date, reason, destination, emergency_phone)


def leave_menu(s):
    s.send(b"Take time off\n")
    s.send(b"Cancel holiday\n")
    s.send(b"Check current bookings\n")
    msg = s.recv(1024).strip(b"\n")
    if not msg:
        return None
    if msg.lower() == b"take time off":
        return create_holiday(s)
    if msg.lower() == b"cancel holiday":
        return cancel_holiday(s)
    if msg.lower() == b"check current bookings":
        return check_holidays(s)


def create_holiday(s):
    s.send(b"Start date?\n")
    start = s.recv(1024).strip(b"\n").decode("utf-8")
    s.send(b"End date?\n")
    end = s.recv(1024).strip(b"\n").decode("utf-8")
    s.send(b"Why do you want to take holiday?\n")
    reason = s.recv(1024).strip(b"\n").decode("utf-8")
    s.send(b"Where will you go?\n")
    destination = s.recv(1024).strip(b"\n").decode("utf-8")
    s.send(b"How can we reach you during your holiday? Please enter your phone number:\n")
    emergency = s.recv(1024).strip(b"\n").decode("utf-8")
    if start and end and reason and destination and emergency:
        start =  actions.user.sanitize(start)
        end = actions.user.sanitize(end)
        reason =  actions.user.sanitize(reason)
        destination =  actions.user.sanitize(destination)
        emergency =  actions.user.sanitize(emergency)
        leave = Leave("2", "0", start, end, reason, destination, emergency)
        return leave
    return None


def cancel_holiday(s):
    s.send(b"Start date?\n")
    start = s.recv(1024).strip(b"\n").decode("utf-8")
    s.send(b"End date?\n")
    end = s.recv(1024).strip(b"\n").decode("utf-8")
    if start and end:
        start =  actions.user.sanitize(start)
        end = actions.user.sanitize(end)
        leave = Leave("2", "1", start, end, " ", " ", " ")
        return leave
    return None


def check_holidays(s):
    start = "01-01-99"
    end = "01-01-99"
    leave = Leave("2", "2", start, end, " ", " ", " ")
    return leave

File: german-telework/service/holiday/Holiday/Leave.cs
using System;
using System.Globalization;


public class Leave {
    public String id;
    public String requestType {get; set;}
    public DateTime start {get; }
    public DateTime end {get; }
    String reason;
    String destination;
    String phone;


    public Leave(String id, String requestType, DateTime start, DateTime end, String reason, String destination, String phone){
        this.id = id;
        this.requestType = requestType;
        this.start = start;
        this.end = end;
        this.reason = reason;
        this.destination = destination;
        this.phone = phone;
    }

    public static Leave Deserialize (String data) {
        String[] parts = data.Split("|");
        String id = parts[0];
        String requestType = parts[1];
        CultureInfo c = new CultureInfo("de-DE");
        DateTime start = new DateTime(1900, 1, 1);
        DateTime end = new DateTime(1900, 1, 1);

        try{
            start = DateTime.Parse(parts[2], c);
            end = DateTime.Parse(parts[3], c);
        } catch (FormatException){
            return null;
        }
        if (start == null || end == null)
            return null;
        String reason = parts[4];
        String destination = parts[5];
        String phone = parts[6];
        return new Leave(id, requestType, start, end, reason, destination, phone); 
    }

    public String Serialize() {
        String serialized = "";
        serialized += id.ToString() + "|";
        serialized += requestType.ToString() + "|";
        CultureInfo c = new CultureInfo("de-DE");
        serialized += start.ToString(c) + "|";
        serialized += end.ToString(c) + "|";
        serialized += reason.ToString() + "|";
        serialized += destination.ToString() + "|";
        serialized += phone.ToString();
        return serialized;
    }

    public static String SerializeList(List<Leave> leaveList) {
        if (leaveList.Count == 0)
            return "";
        String serialized = "";
        foreach (Leave l in leaveList){
            serialized += l.Serialize() + "||";
        }
        return serialized.Remove(serialized.Length - 2);
    }

    public int ComputeTime () {
        return (this.end - this.start).Days;
    }
}
